File: lib/platform/darwin/engine.mjs
/**
 * The macOS resident capture engine, and the build step it needs.
 *
 * `darwin.mjs` captures through `/usr/sbin/screencapture`, which is the engine of
 * record: it is part of macOS, it has no build step, and calling it cannot
 * invalidate anything. What it cannot do is start quickly. Measured on a 4K
 * panel, one call costs about 45ms of process start plus the encoding, so a
 * change check costs 56-90ms and a burst cannot take its first frame until
 * 150-200ms into an animation — two thirds of a 300ms transition gone before the
 * picture is looked at.
 *
 * So there is a second engine, and it is an optimisation with a lifecycle rather
 * than a replacement. `engine.swift` is compiled once into a cache directory,
 * started once, kept resident while it is being used, and dropped when it is not.
 * Every failure path in this file ends the same way — with the caller falling
 * back to `screencapture` — so the worst case is the speed the plugin had before
 * the engine existed, never a capture that does not happen.
 *
 * ## Why the source is compiled here rather than shipped as a binary
 *
 * Apple provides no interpreter for ScreenCaptureKit: `CGDisplayCreateImage` is
 * obsoleted in macOS 15, so the framework is the only supported way to read the
 * screen, and it can only be reached from a compiled program. The alternatives
 * were a prebuilt binary in the repository — which pins the plugin to the SDK it
 * was built against, on a platform whose capture API has already been obsoleted
 * once — or compiling from source on the machine that will run it.
 *
 * Compiling is the honest option, and it is cheap: a second or two, once, on the
 * first capture. It happens *after* the first capture has already been served by
 * `screencapture`, in the background, so no call ever waits for it. A machine
 * without the developer tools simply never gets a helper and keeps using
 * `screencapture`, which is why the fallback is not an error path but the
 * baseline.
 *
 * ## What it does not do
 *
 * The helper knows how to read a rectangle and how to fingerprint one. It does
 * not know about modes, flags, validation, retention or when a burst should
 * stop; all of that stays in the JavaScript, so there is exactly one place each
 * of those lives and the two engines cannot drift apart in what they mean.
 *
 * @module dsh-screen-eye/platform/darwin/engine
 */

import { spawn } from 'node:child_process';
import { createHash } from 'node:crypto';
import { existsSync } from 'node:fs';
import { mkdir, readFile, rename, stat } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { CaptureError } from '../../capture-error.mjs';

/**
 * The helper's source, as a filesystem path.
 *
 * `fileURLToPath` rather than `URL.pathname`: the latter leaves the URL's
 * percent-encoding in place, so a checkout in a directory with a space in its
 * name — which is where this was developed — handed `swiftc` a path containing
 * `%20` and it reported the file missing. The failure is invisible on a checkout
 * with no spaces in its path, which is exactly the kind of bug that reaches a
 * user before it reaches a test.
 */
const ENGINE_SOURCE_PATH = fileURLToPath(new URL('./engine.swift', import.meta.url));

/** The `kind` a failure carries when the engine could not run at all. */
export const ENGINE_UNAVAILABLE = 'engine-unavailable';

/** How long an idle engine is kept before it is let go. */
const ENGINE_IDLE_MS = 120000;

/** How long to wait for a freshly started engine to announce itself. */
const ENGINE_START_TIMEOUT_MS = 20000;

/** How long to wait for a reply before deciding the engine has wedged. */
const ENGINE_REQUEST_TIMEOUT_MS = 120000;

/** How long the compile is given before it is abandoned as too slow. */
const COMPILE_TIMEOUT_MS = 120000;

/** The compiled helper's own source, read once. */
let sourceCache = null;

/**
 * The helper's Swift source.
 * @returns the source text.
 */
export async function engineSource() {
  sourceCache ??= await readFile(new URL('./engine.swift', import.meta.url), 'utf8');
  return sourceCache;
}

/**
 * Where the compiled helper for a given source lives.
 *
 * The name carries a hash of the source, so a changed helper is a different file
 * rather than one that overwrites a binary another process may be running — and
 * so a stale build can never be mistaken for a current one.
 * @param source - the Swift source it will be built from.
 * @returns the absolute path.
 */
export function engineBinaryPath(source) {
  const digest = createHash('sha256').update(source).digest('hex').slice(0, 16);
  return join(tmpdir(), 'dsh-screen-eye-engine', `engine-${digest}`);
}

/**
 * The compiler, or `null` where there is none.
 *
 * `xcrun` is preferred over a bare `swiftc` because it resolves the toolchain
 * the active developer directory selects, which is what a machine with more than
 * one Xcode installed expects. A Command Line Tools install satisfies it too,
 * which matters: the full Xcode is not needed to build this.
 * @returns the command to run, or `null`.
 */
export function compilerCommand() {
  for (const candidate of ['/usr/bin/xcrun', '/usr/bin/swiftc']) {
    if (existsSync(candidate)) return candidate;
  }
  return null;
}

/**
 * Build the helper if it is not already built.
 *
 * Returns `null` rather than throwing when the helper cannot be built, because
 * every caller treats that as "use `screencapture`" and none of them has
 * anything better to do about it. The reason is kept for `screen_permission`,
 * which is where a user would look.
 *
 * @returns the binary's path, or `null`.
 */
export async function buildEngine() {
  const source = await engineSource();
  const binary = engineBinaryPath(source);
  if (existsSync(binary)) return binary;

  const compiler = compilerCommand();
  if (compiler === null) return null;

  await mkdir(dirname(binary), { recursive: true });
  // Compiled to a temporary name and moved into place, so two processes racing
  // to build the same helper cannot leave a half-written binary where a working
  // one belongs. `rename` within a directory is atomic, which is the whole point.
  const staging = `${binary}.${process.pid}.building`;
  const args = compiler.endsWith('xcrun')
    ? ['swiftc', '-O', '-o', staging, ENGINE_SOURCE_PATH]
    : ['-O', '-o', staging, ENGINE_SOURCE_PATH];

  await new Promise((resolve, reject) => {
    const child = spawn(compiler, args, { stdio: ['ignore', 'ignore', 'pipe'] });
    let stderr = '';
    child.stderr.on('data', (chunk) => { stderr += chunk; });
    const timer = setTimeout(() => {
      try { child.kill(); } catch { /* already gone */ }
      reject(new Error(`the helper did not compile within ${COMPILE_TIMEOUT_MS}ms`));
    }, COMPILE_TIMEOUT_MS);
    child.on('error', (error) => { clearTimeout(timer); reject(error); });
    child.on('exit', (code) => {
      clearTimeout(timer);
      if (code === 0) resolve();
      else reject(new Error(stderr.trim().slice(0, 500) || `swiftc exited with code ${code}`));
    });
  });

  const { rename } = await import('node:fs/promises');
  await rename(staging, binary);
  return binary;
}

/** The engine record for the process, or `null` when none is running. */
let engine = null;

/**
 * Start the engine, or return the one already running.
 *
 * Concurrent callers share one start, and a failed start is not cached: the
 * reason for a failure is usually a moment rather than a condition, and giving
 * up permanently would turn one bad start into a plugin that never uses the
 * helper again for the life of the process.
 *
 * @returns the engine record once it has announced its readiness.
 */
function ensureEngine() {
  if (engine !== null && engine.ready && engine.child.exitCode === null && !engine.child.killed) {
    return Promise.resolve(engine);
  }
  if (engine?.starting !== undefined) return engine.starting;
  const record = { child: null, pending: new Map(), buffer: '', nextId: 1, idleTimer: null, ready: false, starting: null, info: null, stderr: '' };
  engine = record;
  record.starting = (async () => {
    const binary = await buildEngine();
    if (binary === null) {
      throw new CaptureError('the macOS capture helper could not be built: no Swift compiler on this machine', {
        kind: ENGINE_UNAVAILABLE,
      });
    }
    const child = spawn(binary, [], { stdio: ['pipe', 'pipe', 'pipe'] });
    record.child = child;
    // Unreferenced, all three streams. A resident process that held the event
    // loop open would keep the *harness* alive, and a harness that will not exit
    // is a worse bug than a slow capture. Letting go costs nothing: when this
    // process ends the engine's stdin closes and it exits by itself, which is
    // the shutdown protocol it was written with.
    child.unref();
    child.stdin.unref?.();
    child.stdout.unref?.();
    child.stderr.unref?.();
    child.stdout.on('data', (chunk) => { consumeEngineOutput(record, chunk); });
    child.stderr.on('data', (chunk) => { record.stderr += chunk; });
    child.on('exit', (code) => {
      const error = new CaptureError(`the macOS engine exited with code ${code}`, {
        kind: ENGINE_UNAVAILABLE,
        detail: record.stderr.trim().slice(0, 500),
      });
      for (const waiter of record.pending.values()) waiter.reject(error);
      record.pending.clear();
      record.ready = false;
      if (engine === record) engine = null;
    });
    await new Promise((resolve, reject) => {
      const timer = setTimeout(() => reject(new CaptureError(`the macOS engine did not start within ${ENGINE_START_TIMEOUT_MS}ms`, { kind: ENGINE_UNAVAILABLE, detail: record.stderr.trim().slice(0, 500) })), ENGINE_START_TIMEOUT_MS);
      record.onReady = () => { clearTimeout(timer); resolve(); };
      record.onFailed = (error) => { clearTimeout(timer); reject(error); };
    });
    return record;
  })().catch((error) => {
    if (engine === record) engine = null;
    try { record.child?.kill(); } catch { /* already gone */ }
    throw error;
  });
  return record.starting;
}

/**
 * Take one line of engine output: the readiness announcement, or a reply.
 * @param record - the engine record.
 * @param chunk - bytes from the engine's stdout.
 */
function consumeEngineOutput(record, chunk) {
  record.buffer += chunk;
  let index;
  while ((index = record.buffer.indexOf('\n')) >= 0) {
    const line = record.buffer.slice(0, index).trim();
    record.buffer = record.buffer.slice(index + 1);
    if (line === '') continue;
    let payload;
    try { payload = JSON.parse(line); } catch { continue; }
    if (payload.ready === true) {
      record.ready = true;
      record.info = payload;
      record.onReady?.();
      continue;
    }
    if (payload.ok === false && payload.id === undefined) {
      // The engine saying it could not set itself up. Nothing is waiting on a
      // reply yet, so this has to reach the start's own promise.
      record.onFailed?.(new CaptureError(payload.detail ?? 'the macOS engine could not start', {
        kind: payload.kind ?? ENGINE_UNAVAILABLE,
        detail: payload.detail ?? '',
      }));
      continue;
    }
    const waiter = record.pending.get(payload.id);
    if (waiter === undefined) continue;
    record.pending.delete(payload.id);
    clearTimeout(waiter.timer);
    waiter.resolve(payload);
  }
}

/**
 * Send one request to the engine and wait for its reply.
 * @param request - the request body; an id is added here.
 * @param options - cancellation signal and budget.
 * @returns the decoded reply.
 */
async function engineRequest(request, options = {}) {
  const record = await ensureEngine();
  const id = record.nextId;
  record.nextId += 1;
  return new Promise((resolve, reject) => {
    const settle = (fn, value) => {
      record.pending.delete(id);
      options.signal?.removeEventListener('abort', onAbort);
      fn(value);
    };
    const onAbort = () => settle(reject, new CaptureError('the engine request was aborted', { kind: 'capture-failed' }));
    if (options.signal?.aborted === true) { onAbort(); return; }
    const timer = setTimeout(() => {
      settle(reject, new CaptureError(`the macOS engine did not answer request ${id} within ${ENGINE_REQUEST_TIMEOUT_MS}ms`, { kind: ENGINE_UNAVAILABLE }));
    }, options.timeoutMs ?? ENGINE_REQUEST_TIMEOUT_MS);
    record.pending.set(id, { resolve: (payload) => settle(resolve, payload), reject: (error) => settle(reject, error), timer });
    options.signal?.addEventListener('abort', onAbort, { once: true });
    record.child.stdin.write(`${JSON.stringify({ ...request, id })}\n`);
    scheduleEngineIdle(record);
  });
}

/**
 * Keep the engine for a while after its last request, then let it go.
 * @param record - the engine record.
 */
function scheduleEngineIdle(record) {
  clearTimeout(record.idleTimer);
  record.idleTimer = setTimeout(() => { if (engine === record) stopEngine(); }, ENGINE_IDLE_MS);
  record.idleTimer.unref?.();
}

/**
 * Whether a resident engine is answering right now.
 *
 * The distinction the callers need is "warm or not", never "does this platform
 * have an engine": a cold engine must not delay a capture, and asking the
 * question must not start one.
 * @returns whether the engine is ready.
 */
export function engineIsWarm() {
  return engine !== null && engine.ready && engine.child.exitCode === null && !engine.child.killed;
}

/**
 * Start the engine without waiting for it, for the call after this one.
 *
 * This is what makes the compile invisible. The first capture is served by
 * `screencapture` — already running, no worse than the plugin has ever been —
 * and the helper is built and started behind it, so the second capture has an
 * engine. A build that fails is recorded and never retried in a way that could
 * cost a call anything.
 */
export function warmEngine() {
  if (engineIsWarm()) return;
  ensureEngine().catch(() => { /* the fallback is the plan, not a failure */ });
}

/**
 * Shut the engine down, waiting briefly for it to leave of its own accord.
 */
export function stopEngine() {
  const record = engine;
  if (record === null) return;
  engine = null;
  clearTimeout(record.idleTimer);
  const error = new CaptureError('the macOS engine was shut down', { kind: ENGINE_UNAVAILABLE });
  for (const waiter of record.pending.values()) waiter.reject(error);
  record.pending.clear();
  // A record exists as soon as a start begins, and a start that never got as far
  // as spawning — a build that failed, or is still running — has no child. That
  // is the state a shutdown during the first capture finds, so it has to be a
  // case rather than a crash.
  if (record.child === null) return;
  try { record.child.stdin.end(); } catch { /* already closed */ }
  const killer = setTimeout(() => { try { record.child.kill(); } catch { /* already gone */ } }, 1000);
  killer.unref?.();
  record.child.once('exit', () => clearTimeout(killer));
}

// A harness that exits takes its engine with it; the engine would also notice on
// its own, because its stdin closes, but this makes it immediate.
process.once('exit', () => {
  try { engine?.child?.kill(); } catch { /* nothing to kill */ }
});

/**
 * Ask the running engine to capture one frame.
 *
 * The rectangle arrives already resolved, as the fields the engine's protocol
 * takes. It is deliberately not resolved here from a plan: the mapping from a
 * plan onto a screen rectangle is the one thing both engines must agree on, so
 * it lives in exactly one function — `engineTarget` in `darwin.mjs` — and is
 * asserted there. Resolving it a second time in this file is how the two would
 * come to disagree, which is the failure this tool can least afford: a plausible
 * picture of the wrong place. An earlier revision did resolve it here, from
 * `plan.origin` and `plan.size`, which a capture plan does not have — every
 * field read `undefined`, every capture took the whole screen, and a region
 * silently returned a 4K image of the entire desktop.
 *
 * @param target - the resolved `{ display, x, y, width, height, cursor }`.
 * @param outputPath - the resolved absolute output path.
 * @param options - cancellation signal and budget.
 * @returns the written PNG's path and byte length.
 * @throws a `CaptureError` when the engine could not answer.
 */
export async function engineCapture(target, outputPath, options = {}) {
  const reply = await engineRequest({ ...target, op: 'capture', path: outputPath }, options);
  if (reply.ok !== true) {
    throw new CaptureError(reply.detail ?? 'the macOS engine refused the capture', {
      kind: reply.kind ?? 'capture-failed',
      detail: reply.detail ?? '',
    });
  }
  let size;
  try {
    size = (await stat(outputPath)).size;
  } catch {
    throw new CaptureError('the engine reported a capture but wrote no file', { kind: 'capture-failed' });
  }
  if (size === 0) throw new CaptureError('the engine wrote an empty file', { kind: 'capture-failed' });
  return { outputPath, bytes: size };
}

/**
 * Fingerprint a rectangle through the running engine.
 *
 * The comparison a wait needs is answered from a 64x64 reduction of the region
 * rather than from a PNG, which is what makes it about 23ms instead of 56ms.
 * @param target - the resolved `{ display, x, y, width, height }`.
 * @param options - cancellation signal.
 * @returns the fingerprint as a string.
 * @throws a `CaptureError` when the engine could not answer.
 */
export async function engineFingerprint(target, options = {}) {
  const reply = await engineRequest({ ...target, op: 'poll' }, options);
  if (reply.ok !== true) {
    throw new CaptureError(reply.detail ?? 'the macOS engine refused the check', {
      kind: reply.kind ?? 'capture-failed',
      detail: reply.detail ?? '',
    });
  }
  return String(reply.hash);
}

/** Reset the module's cached source, for tests that change it. */
export function forgetEngineSource() {
  sourceCache = null;
}

File: lib/platform/darwin/engine.swift
// The macOS capture engine: set ScreenCaptureKit up once, then answer one
// request per line of stdin.
//
// This exists for one reason, and it is not speed in general. A one-shot
// animation — a hover, a panel opening, a page load — is over in a few hundred
// milliseconds, and a watcher cannot start observing it before its engine
// exists. `/usr/sbin/screencapture` costs about 45ms of process start whatever
// it captures, so every change check pays it, and a wait that confirms a change
// twice pays it twice: measured, a burst issued at the moment an animation began
// landed its first frame 150-200ms in, by which time two thirds of a 300ms
// transition had already happened. Resident, the same rectangle is read in about
// 13ms and hashed in about 23ms, so the same burst starts inside the animation
// instead of after it.
//
// It is deliberately *not* a second implementation of the plugin. All it does is
// read a rectangle and answer what it looks like; the flags, the modes, the
// validation, the retention and the decision about when to stop all stay in the
// JavaScript, so there is one place where each of those lives. `screencapture`
// remains the engine of record — it is the thing with no build step and no
// version to match — and this process is an optimisation that can fail, at which
// point the caller falls back to it.
//
// Protocol: one JSON object per line in, one JSON object per line out, each
// reply carrying the request's own `id` so a caller may have several
// outstanding. Closing stdin is the shutdown protocol, which is what cleans up
// after a harness that dies without saying goodbye.

import Foundation
import ScreenCaptureKit
import CoreGraphics
import ImageIO
import UniformTypeIdentifiers

// MARK: - Output

/// Write one line and flush it. A reply that sits in a buffer is a caller that
/// waits for a timeout it did not need to wait for.
func emit(_ text: String) {
    FileHandle.standardOutput.write((text + "\n").data(using: .utf8)!)
}

func emitJSON(_ object: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: object),
          let text = String(data: data, encoding: .utf8) else {
        emit(#"{"ok":false,"detail":"could not encode a reply"}"#)
        return
    }
    emit(text)
}

func fail(_ id: Int, _ kind: String, _ detail: String) {
    emitJSON(["id": id, "ok": false, "kind": kind, "detail": detail])
}

// MARK: - Requests

struct Request {
    let id: Int
    let op: String
    let display: Int
    let x: Int
    let y: Int
    let width: Int
    let height: Int
    let path: String
    let cursor: Bool

    init?(_ line: String) {
        guard let data = line.data(using: .utf8),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        id = raw["id"] as? Int ?? 0
        op = raw["op"] as? String ?? "capture"
        display = raw["display"] as? Int ?? 1
        x = raw["x"] as? Int ?? 0
        y = raw["y"] as? Int ?? 0
        width = raw["width"] as? Int ?? 0
        height = raw["height"] as? Int ?? 0
        path = raw["path"] as? String ?? ""
        cursor = raw["cursor"] as? Bool ?? false
    }
}

// MARK: - Capturing

/// The screen rectangle a request names, in the display's own coordinate space.
///
/// Requests carry a rectangle in the desktop's global space, which is the space
/// `region` is documented in and the space `screencapture -R` takes. A display
/// that does not start at the desktop's origin — a second screen placed to the
/// left or above the main one — has its own origin, and `sourceRect` is measured
/// from *that*, so the display's origin is subtracted here. Getting this wrong
/// captures the right size of the wrong place, which for a tool whose whole
/// purpose is reading the screen is the worst possible failure: a plausible
/// picture of something else.
///
/// The subtraction is load-bearing and was verified the hard way on a second
/// display. A request arrives in the desktop's global coordinates — the space
/// `region` is documented in — while `sourceRect` is measured from the display's
/// own top-left corner, so the display's frame origin comes off. On the main
/// display `frame.min` is (0, 0) and every one of these forms behaves the same,
/// which is exactly why the mistake survives on a single-screen machine: with a
/// Sidecar display at frame (-748, 2160), subtracting both is the only form that
/// captures anything at all, and passing the global rectangle through unchanged
/// fails with `SCStreamErrorDomain Code=-3812`.
func sourceRect(_ request: Request, _ display: SCDisplay) -> CGRect? {
    if request.width > 0 && request.height > 0 {
        return CGRect(x: CGFloat(request.x) - display.frame.minX,
                      y: CGFloat(request.y) - display.frame.minY,
                      width: CGFloat(request.width),
                      height: CGFloat(request.height))
    }
    return nil
}

/// How many native pixels this display has per point.
///
/// This is the difference between a correct capture and a broken one on any
/// display that is not 1x, and it was found the hard way — on a 2x Sidecar
/// display the engine returned **half** the panel's resolution for a full-screen
/// capture, and refused a region outright with
/// `SCStreamErrorDomain Code=-3812 "the operation could not be completed"`.
///
/// The cause is that ScreenCaptureKit speaks in *points* while the display has
/// *pixels*: `SCDisplay.width` is the point width (1194 for a 2388-wide iPad)
/// and wants to be told the output size in pixels. Asking for 1194x834 asked the
/// framework for a half-resolution copy, and pairing a point-sized `sourceRect`
/// with it produced the invalid-parameter error on a region.
///
/// The scale comes from the display *mode*, and specifically from
/// `pixelWidth / width`. That pairing was measured, not assumed: on the same
/// machine and in the same process, `CGDisplayPixelsWide` reports 1194 for the
/// iPad — the point width, not the pixel width — so the obvious
/// `CGDisplayPixelsWide / CGDisplayBounds.width` ratio comes out as 1 and leaves
/// the bug in place. Only the mode returns both numbers, and only their ratio is
/// right.
///
/// @param display - the display to measure.
/// @returns the scale factor, 1 when it cannot be determined.
func backingScale(_ display: SCDisplay) -> CGFloat {
    guard let mode = CGDisplayCopyDisplayMode(display.displayID), mode.width > 0 else { return 1 }
    let scale = CGFloat(mode.pixelWidth) / CGFloat(mode.width)
    return scale > 0 ? scale : 1
}

/// Read a rectangle, or the whole display when no size was asked for.
func grab(_ request: Request, _ display: SCDisplay) async throws -> CGImage {
    let filter = SCContentFilter(display: display, excludingWindows: [])
    let config = SCStreamConfiguration()
    let scale = backingScale(display)
    if let rect = sourceRect(request, display) {
        // The rectangle is in points, in the display's own space, and
        // `sourceRect` is documented in points too — so only the *output* size
        // is scaled. That is the asymmetry the error above came from: the
        // request was consistent, the answer was not.
        config.sourceRect = rect
        config.width = Int(CGFloat(request.width) * scale)
        config.height = Int(CGFloat(request.height) * scale)
    } else {
        // "The whole display" means its native resolution, not its point size.
        config.width = Int(CGFloat(display.width) * scale)
        config.height = Int(CGFloat(display.height) * scale)
    }
    config.showsCursor = request.cursor
    config.captureResolution = .best
    return try await SCScreenshotManager.captureImage(contentFilter: filter, configuration: config)
}

/// Write a capture as PNG.
func write(_ image: CGImage, to path: String) -> Bool {
    guard let destination = CGImageDestinationCreateWithURL(
        URL(fileURLWithPath: path) as CFURL, UTType.png.identifier as CFString, 1, nil) else { return false }
    CGImageDestinationAddImage(destination, image, nil)
    return CGImageDestinationFinalize(destination)
}

/// A cheap fingerprint of what a rectangle looks like right now.
///
/// The comparison a wait needs is "same or different", and the honest way to
/// answer it does not require a PNG: the rectangle is drawn into a 64x64 bitmap
/// and those pixels are hashed in process. That is about 23ms against the 56ms
/// a `screencapture` costs, and unlike the file comparison it cannot be
/// disturbed by anything the encoder writes beside the pixels.
///
/// The cost of the shortcut is the same one Windows' sampling accepts: a change
/// too small to survive a 64x64 reduction is not seen. That is why the shared
/// wait requires a change to be confirmed twice rather than believing one
/// crossing, and why its guidance tells the caller to watch the region the
/// animation is in rather than the whole desktop.
func fingerprint(_ request: Request, _ display: SCDisplay, into buffer: inout [UInt8], size: Int) throws -> UInt64 {
    let image = try awaitSync { try await grab(request, display) }
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let context = CGContext(data: &buffer, width: size, height: size, bitsPerComponent: 8,
                                  bytesPerRow: size * 4, space: colorSpace,
                                  bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else {
        throw NSError(domain: "screen-eye", code: 1, userInfo: [NSLocalizedDescriptionKey: "no bitmap context"])
    }
    context.interpolationQuality = .low
    context.draw(image, in: CGRect(x: 0, y: 0, width: size, height: size))
    var hash: UInt64 = 0xcbf2_9ce4_8422_2325      // FNV-1a, for no reason beyond being short and stable
    for byte in buffer {
        hash ^= UInt64(byte)
        hash = hash &* 0x0000_0100_0000_01b3
    }
    return hash
}

/// Run an async operation from the synchronous request loop.
///
/// The loop is synchronous on purpose — a resident engine that answered out of
/// order would be a resident engine whose replies cannot be matched to its
/// requests by the `id` alone — so each request blocks its thread on a semaphore
/// while the work happens on a Task.
func awaitSync<T>(_ work: @escaping () async throws -> T) throws -> T {
    let semaphore = DispatchSemaphore(value: 0)
    var result: Result<T, Error>?
    Task {
        do { result = .success(try await work()) } catch { result = .failure(error) }
        semaphore.signal()
    }
    semaphore.wait()
    return try result!.get()
}

// MARK: - The loop

let setupStarted = Date()
let content: SCShareableContent
do {
    content = try awaitSync { try await SCShareableContent.excludingDesktopWindows(false, onScreenWindowsOnly: true) }
} catch {
    // The usual cause is a missing Screen Recording grant. Saying so here lets
    // the caller fall back to `screencapture`, which reports the same condition
    // in the words the plugin already knows how to turn into onboarding steps.
    emitJSON(["ok": false, "kind": "engine-unavailable", "detail": "\(error)"])
    exit(1)
}
let displays = content.displays
let setupMs = Int(Date().timeIntervalSince(setupStarted) * 1000)

// The readiness line carries the inventory, so the caller does not have to ask
// a second question to learn what this process can see.
emitJSON([
    "ok": true,
    "ready": true,
    "setupMs": setupMs,
    "displays": displays.map { ["index": 0, "width": $0.width, "height": $0.height,
                                "x": Int($0.frame.minX), "y": Int($0.frame.minY)] },
])

var buffer = [UInt8](repeating: 0, count: 64 * 64 * 4)

while let line = readLine(strippingNewline: true) {
    if line.isEmpty { continue }
    guard let request = Request(line) else {
        // A line that is not a request is not fatal: the engine stays up and the
        // next line is read, because a caller that sends one malformed request
        // should not lose the process it just paid to start.
        emitJSON(["ok": false, "kind": "bad-request", "detail": "not a JSON request"])
        continue
    }
    let index = request.display - 1
    guard index >= 0 && index < displays.count else {
        fail(request.id, "display-missing", "display \(request.display) does not exist: this machine reports \(displays.count) display(s)")
        continue
    }
    let display = displays[index]

    do {
        switch request.op {
        case "poll":
            let started = Date()
            let hash = try fingerprint(request, display, into: &buffer, size: 64)
            emitJSON(["id": request.id, "ok": true, "hash": String(hash),
                      "ms": Int(Date().timeIntervalSince(started) * 1000)])
        case "capture":
            guard !request.path.isEmpty else { fail(request.id, "bad-request", "capture needs a path"); continue }
            let started = Date()
            let image = try awaitSync { try await grab(request, display) }
            guard write(image, to: request.path) else {
                fail(request.id, "capture-failed", "could not write \(request.path)"); continue
            }
            emitJSON(["id": request.id, "ok": true, "width": image.width, "height": image.height,
                      "ms": Int(Date().timeIntervalSince(started) * 1000)])
        default:
            fail(request.id, "bad-request", "unknown op \(request.op)")
        }
    } catch {
        fail(request.id, "capture-failed", "\(error)")
    }
}
